src/map.rs

```diff
@@ -1186,18 +1191,24 @@ fn handle_underflow<Lifetime, K, V>(mut cur_node: NodeRef<Lifetime, K, V, marker
                     Ok(mut right) => {
                         // steal or merge right
                         if right.can_merge() {
-                            cur_node = right.merge().into_node().forget_type();
+                            match right.merge() {
+                                Some(node) => cur_node = node.into_node().forget_type(),
+                                None => return,
+                            }
                         } else {
-                            let (k, v) = right.reborrow_mut().right_edge().descend().remove(0); // TODO: Reuse cur_node?
-                            let k = mem::replace(right.reborrow_mut().into_kv().0, k);
-                            let v = mem::replace(right.reborrow_mut().into_kv().1, v);
-                            right.left_edge().descend().push(k, v);
+                            let (k, v, edge) = right.reborrow_mut().right_edge().descend().pop_front(); // TODO: Reuse cur_node?
+                            let k = mem::replace(right.reborrow_mut().into_kv_mut().0, k);
+                            let v = mem::replace(right.reborrow_mut().into_kv_mut().1, v);
+                            match right.reborrow_mut().left_edge().descend().force() {
+                                Leaf(mut leaf) => leaf.push(k, v),
+                                Internal(mut internal) => internal.push(k, v, edge.unwrap())
+                            }
                             cur_node = right.into_node().forget_type();
                         }
                     },
-                    Err(parent) => {
+                    Err(mut parent) => {
                         // The parent node is underfull, so we must be at the root.
-                        parent.into_node().reborrow_mut().into_root_mut().shrink();
+                        parent.reborrow_mut().into_node().into_root_mut().shrink();
```

Comment:
This is actually really dangerous, as this function is marked as taking a `NodeRef` of any lifetime, but when the lifetime is `Owned`, `into_root_mut` isn't defined, as the root pointers aren't set. Thus, if an owned node were to be passed to `handle_underflow`, this would at best segfault. I think I'd prefer to just skip the `reborrow_mut` and force the input to `handle_underflow` to be `Borrowed`.